nixos/modules/services/networking/unbound.nix

```diff
@@ -6,41 +6,113 @@ let
 
   cfg = config.services.unbound;
 
-  stateDir = "/var/lib/unbound";
-
-  access = concatMapStringsSep "\n  " (x: "access-control: ${x} allow") cfg.allowedAccess;
-
-  interfaces = concatMapStringsSep "\n  " (x: "interface: ${x}") cfg.interfaces;
-
-  isLocalAddress = x: substring 0 3 x == "::1" || substring 0 9 x == "127.0.0.1";
-
-  forward =
-    optionalString (any isLocalAddress cfg.forwardAddresses) ''
-      do-not-query-localhost: no
-    '' +
-    optionalString (cfg.forwardAddresses != []) ''
-      forward-zone:
-        name: .
-    '' +
-    concatMapStringsSep "\n" (x: "    forward-addr: ${x}") cfg.forwardAddresses;
-
-  rootTrustAnchorFile = "${stateDir}/root.key";
-
-  trustAnchor = optionalString cfg.enableRootTrustAnchor
-    "auto-trust-anchor-file: ${rootTrustAnchorFile}";
-
-  confFile = pkgs.writeText "unbound.conf" ''
-    server:
-      directory: "${stateDir}"
-      username: unbound
-      chroot: "${stateDir}"
-      pidfile: ""
-      ${interfaces}
-      ${access}
-      ${trustAnchor}
-    ${cfg.extraConfig}
-    ${forward}
-  '';
+  settingsSubmodule = let
+    validConfigTypes = with types; either int (either str (either bool float));
+  in {
+    options = {
+      server = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = {
```

Comment:
I don't think that this is a good solution here: when I e.g. define `settings.server.interface`, the `server:`-block will only have the `interface`-value, anything else is gone since this default-value has a lower priority than my own `settings.server`-declaration.

Instead, I'd recommend to do sth. like this in the module:

``` nix
{
  settings.server = {
    directory = mkDefault cfg.stateDir;
    /* and so on */
  };
}
```

With this approach, it's possible to define `settings` multiple times without loosing all defaults.

The default value for `directory` can be overriden in a config like this: `settings.server.directory = "/var/lib/alt-dir";`.

Reply:
This is kind of a duplicate of another comment.